compiler/rustc_infer/src/infer/outlives/obligations.rs

```diff
@@ -324,17 +327,57 @@ where
         self.delegate.push_verify(origin, generic, region, verify_bound);
     }
 
+    #[instrument(level = "debug", skip(self))]
+    fn opaque_must_outlive(
+        &mut self,
+        def_id: DefId,
+        substs: SubstsRef<'tcx>,
+        origin: infer::SubregionOrigin<'tcx>,
+        region: ty::Region<'tcx>,
+    ) {
+        self.generic_must_outlive(
+            origin,
+            region,
+            GenericKind::Opaque(def_id, substs),
+            def_id,
+            substs,
+            |ty| match *ty.kind() {
+                ty::Opaque(def_id, substs) => (def_id, substs),
+                _ => bug!("expected only projection types from env, not {:?}", ty),
+            },
+        );
+    }
+
+    #[instrument(level = "debug", skip(self))]
     fn projection_must_outlive(
         &mut self,
         origin: infer::SubregionOrigin<'tcx>,
         region: ty::Region<'tcx>,
         projection_ty: ty::ProjectionTy<'tcx>,
     ) {
-        debug!(
-            "projection_must_outlive(region={:?}, projection_ty={:?}, origin={:?})",
-            region, projection_ty, origin
+        self.generic_must_outlive(
+            origin,
+            region,
+            GenericKind::Projection(projection_ty),
+            projection_ty.item_def_id,
+            projection_ty.substs,
+            |ty| match ty.kind() {
+                ty::Projection(projection_ty) => (projection_ty.item_def_id, projection_ty.substs),
+                _ => bug!("expected only projection types from env, not {:?}", ty),
+            },
         );
+    }
 
+    #[instrument(level = "debug", skip(self, filter))]
+    fn generic_must_outlive(
+        &mut self,
+        origin: infer::SubregionOrigin<'tcx>,
+        region: ty::Region<'tcx>,
+        generic: GenericKind<'tcx>,
+        def_id: DefId,
+        substs: SubstsRef<'tcx>,
+        filter: impl Fn(Ty<'tcx>) -> (DefId, SubstsRef<'tcx>),
+    ) {
```

Comment:
Yea, I was worried about performance here, which is why I wanted to get a baseline perf run with the current version, the problem is that I broke a stable crate :(

```rust
use std::pin::Pin;
use std::future::Future;

pub struct ReusableBoxFuture<'a, T> {
    boxed: Pin<Box<dyn Future<Output = T> + 'a>>,
}

impl<'a, T> ReusableBoxFuture<'a, T> {
    pub fn new<F>(future: F) -> Self
    where
        F: Future<Output = T> + 'a,
    {
        unimplemented!()
    }
}

fn new<T>() -> ReusableBoxFuture<'static, T> {
    ReusableBoxFuture::new(async { unreachable!() })
}

fn main() {
}
```

errors with

```
error[E0310]: the opaque type `from_generator<[static generator@/home/ubuntu/rust4/src/test/ui/type-alias-impl-trait/implied_bounds3.rs:22:34: 22:52]>::{opaque#0}` may not live long enough
  --> /home/ubuntu/rust4/src/test/ui/type-alias-impl-trait/implied_bounds3.rs:22:5
   |
LL |     ReusableBoxFuture::new(async { unreachable!() })
   |     ^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider adding an explicit lifetime bound `from_generator<[static generator@/home/ubuntu/rust4/src/test/ui/type-alias-impl-trait/implied_bounds3.rs:22:34: 22:52]>::{opaque#0}: 'static`...
   = note: ...so that the type `impl Future<Output = T>` will meet its required lifetime bounds...
note: ...that is required by this bound
  --> /home/ubuntu/rust4/src/test/ui/type-alias-impl-trait/implied_bounds3.rs:15:33
   |
LL |         F: Future<Output = T> + 'a,
   |    
```

Reply:
Minimized to

```rust
fn foo<F>(_: F)
where
    F: 'static,
{
}

fn from<F: Send>(f: F) -> impl Send {
    f
}

fn bar<T>() {
    foo(from(|| ()))
}
```

giving the following error

```
error[E0310]: the opaque type `from<[closure@/home/ubuntu/rust4/src/test/ui/type-alias-impl-trait/implied_bounds3.rs:16:14: 16:19]>::{opaque#0}` may not live long enough
  --> /home/ubuntu/rust4/src/test/ui/type-alias-impl-trait/implied_bounds3.rs:16:5
   |
LL |     foo(from(|| ()))
   |     ^^^^^^^^^^^^^^^^
   |
   = help: consider adding an explicit lifetime bound `from<[closure@/home/ubuntu/rust4/src/test/ui/type-alias-impl-trait/implied_bounds3.rs:16:14: 16:19]>::{opaque#0}: 'static`...
   = note: ...so that the type `impl Send` will meet its required lifetime bounds
```

This is because with this PR we don't consider the fact that calling `from` with `F` being the type of the `|| ()` closure, we should be able to know that the return type of `from` must outlive `'static` because `F` outlives `'static`. But we explicitly avoid inferring this kind of information, as it is unsound to do.

Now... the above snippet works on stable, so we shouldn't break it, especially since it *is* sound for return position impl trait. It's just that the very same thing would be unsound for return position impl trait.

Changing the example to

```rust
fn from<'a, F: Send + 'a>(f: F) -> impl Send + 'a {
    f
}
```

makes it compile again right now, since now we can sanely infer that the return type outlives `'static` just by knowing `F: 'static`.

Now... I'm going to add some logic to allow return position impl trait to look at its generic parameters to imply lifetime bounds that hold, just so we have back compat. But we need to look into it more to figure out whether this divergence between RPIT and TAIT is desirable.

Reply:
I believe #97406 is sufficient to fix this regression. Am I missing something?

Can you experiment with merging #97406 here, just to get perf run and perhaps another crater run?

Reply:
yes that did the trick! thanks!